# Tuning KNeighborsRegressor

In [2]:
import numpy
import csv
import time
import pandas
import signal
import warnings
warnings.filterwarnings("ignore")

In [3]:
from sklearn.metrics import make_scorer
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_validate
from sklearn.pipeline import Pipeline

In [4]:
from sklearn.multioutput import ClassifierChain
from sklearn.multioutput import MultiOutputRegressor
from sklearn.multioutput import MultiOutputClassifier

from sklearn.ensemble.weight_boosting import AdaBoostClassifier
from sklearn.ensemble.weight_boosting import AdaBoostRegressor
from sklearn.ensemble.bagging import BaggingClassifier
from sklearn.ensemble.bagging import BaggingRegressor
from sklearn.ensemble.forest import ExtraTreesClassifier
from sklearn.ensemble.forest import ExtraTreesRegressor
from sklearn.ensemble.gradient_boosting import GradientBoostingClassifier
from sklearn.ensemble.gradient_boosting import GradientBoostingRegressor
from sklearn.ensemble.forest import RandomForestClassifier
from sklearn.ensemble.forest import RandomForestRegressor

from sklearn.linear_model.bayes import ARDRegression
from sklearn.linear_model.bayes import BayesianRidge
from sklearn.naive_bayes import BernoulliNB
from sklearn.cross_decomposition.cca_ import CCA
from sklearn.tree.tree import DecisionTreeClassifier
from sklearn.tree.tree import DecisionTreeRegressor
from sklearn.linear_model.coordinate_descent import ElasticNet
from sklearn.tree.tree import ExtraTreeClassifier
from sklearn.tree.tree import ExtraTreeRegressor
from sklearn.naive_bayes import GaussianNB
from sklearn.gaussian_process.gpc import GaussianProcessClassifier
from sklearn.gaussian_process.gpr import GaussianProcessRegressor
from sklearn.linear_model.huber import HuberRegressor
from sklearn.neighbors.classification import KNeighborsClassifier
from sklearn.neighbors.regression import KNeighborsRegressor
from sklearn.kernel_ridge import KernelRidge
from sklearn.semi_supervised.label_propagation import LabelPropagation
from sklearn.semi_supervised.label_propagation import LabelSpreading
from sklearn.linear_model.least_angle import Lars
from sklearn.linear_model.coordinate_descent import Lasso
from sklearn.linear_model.least_angle import LassoLars
from sklearn.linear_model.least_angle import LassoLarsIC
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.linear_model.base import LinearRegression
from sklearn.svm.classes import LinearSVC
from sklearn.svm.classes import LinearSVR
from sklearn.linear_model.logistic import LogisticRegression
from sklearn.neural_network.multilayer_perceptron import MLPClassifier
from sklearn.neural_network.multilayer_perceptron import MLPRegressor
from sklearn.linear_model.coordinate_descent import MultiTaskElasticNet
from sklearn.linear_model.coordinate_descent import MultiTaskLasso
from sklearn.naive_bayes import MultinomialNB
from sklearn.neighbors.nearest_centroid import NearestCentroid
from sklearn.svm.classes import NuSVC
from sklearn.svm.classes import NuSVR
from sklearn.linear_model.omp import OrthogonalMatchingPursuit
from sklearn.cross_decomposition.pls_ import PLSCanonical
from sklearn.cross_decomposition.pls_ import PLSRegression
from sklearn.linear_model.passive_aggressive import PassiveAggressiveClassifier
from sklearn.linear_model.passive_aggressive import PassiveAggressiveRegressor
from sklearn.linear_model.perceptron import Perceptron
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.linear_model.ransac import RANSACRegressor
from sklearn.neighbors.classification import RadiusNeighborsClassifier
from sklearn.neighbors.regression import RadiusNeighborsRegressor
from sklearn.linear_model.ridge import Ridge
from sklearn.linear_model.ridge import RidgeClassifier
from sklearn.linear_model.stochastic_gradient import SGDClassifier
from sklearn.linear_model.stochastic_gradient import SGDRegressor
from sklearn.svm.classes import SVC
from sklearn.svm.classes import SVR
from sklearn.linear_model.theil_sen import TheilSenRegressor

In [5]:
mlts = [MultiOutputRegressor, ClassifierChain]
bss = [Ridge, LinearSVR, GaussianProcessRegressor, PassiveAggressiveRegressor, KNeighborsRegressor, MLPRegressor, HuberRegressor, SGDRegressor, LassoLarsIC, ExtraTreeRegressor, ElasticNet, DecisionTreeRegressor, Perceptron, OrthogonalMatchingPursuit, BernoulliNB]

In [6]:
tfidf = pandas.read_csv('07 NM PT TFIDF.csv', delimiter=',', encoding='latin-1').fillna('')
tfidf[:3]

____________________  _____________________  _____________________ follow  \
0                   0.0                    0.0                           0.0   
1                   0.0                    0.0                           0.0   
2                   0.0                    0.0                           0.0   

   able  acceptable  accessory  accessory require  accidentally  accommodate  \
0   0.0         0.0        0.0                0.0           0.0          0.0   
1   0.0         0.0        0.0                0.0           0.0          0.0   
2   0.0         0.0        0.0                0.0           0.0          0.0   

   accompany   ...    work  would  write  writer  wrong  wrong patient  \
0        0.0   ...     0.0    0.0    0.0     0.0    0.0            0.0   
1        0.0   ...     0.0    0.0    0.0     0.0    0.0            0.0   
2        0.0   ...     0.0    0.0    0.0     0.0    0.0            0.0   

   wrong tattoo  xrt  xrt date  xrt pt  
0           0.0  0.0       0.0     0.0  
1           0.0  0.0       0.0     0.0  
2           0.0  0.0       0.0     0.0  

[3 rows x 1891 columns]

In [7]:
ohe = pandas.read_csv('07 NM PT OHE.csv', delimiter=',', encoding='latin-1').fillna('')
ohe[:3]

Wrong target or OAR contours  \
0                             0   
1                             0   
2                             0   

   Treatment plan acceptable but not physically deliverable  \
0                                                  0          
1                                                  0          
2                                                  0          

   Excess imaging dose   \
0                     0   
1                     0   
2                     0   

   Failure to perform on-treatment imaging as per instructions  Wrong patient  \
0                                                  0                        0   
1                                                  0                        0   
2                                                  0                        0   

   Fall or other patient injury or medical condition  \
0                                                  0   
1                                                  0   
2                                                  0   

   Wrong anatomical site (excluding laterality)  \
0                                             0   
1                                             0   
2                                             0   

   Treatment plan (isodose distribution) unacceptable  \
0                                                  0    
1                                                  0    
2                                                  0    

   Systematic hardware/software (including dose-volume) error  \
0                                                  0            
1                                                  0            
2                                                  0            

   Interventional procedure error (Retired value)        ...          \
0                                               0        ...           
1                                               0        ...           
2                                               0        ...           

   Inappropriate or poorly informed decision to treat or plan  \
0                                                  0            
1                                                  0            
2                                                  0            

   Inadequate coordination of combined modality care  \
0                                                  0   
1                                                  0   
2                                                  0   

   Wrong, missing, mislabeled, or damaged treatment accessories  \
0                                                  1              
1                                                  0              
2                                                  0              

   Wrong side (laterality)  \
0                        0   
1                        0   
2                        0   

   Wrong prescription dose-fractionation or calculation error  \
0                                                  0            
1                                                  0            
2                                                  0            

   Radiation therapy scheduling error  \
0                                   0   
1                                   1   
2                                   1   

   Treatment not delivered - personnel/hardware/software failure  \
0                                                  0               
1                                                  0               
2                                                  0               

   Wrong patient position, setup point, or shift  Wrong planning margins  \
0                                              0                       0   
1                                              0                       0   
2                                              0                       0   

   Allergic reaction  
0                  0  
1                  0  
2                

In [8]:
def lowest_correct(trues, preds):
    num_of_options = len(trues)
    drop_down_options = list(reversed(numpy.argsort(preds)))
    correct_options = [i for i in range(num_of_options) if trues[i]==1]
    return max([drop_down_options.index(correct_option) for correct_option in correct_options]) + 1
def average_lowest_correct(list_of_trues, list_of_preds):
    length = len(list_of_trues)
    return numpy.mean([lowest_correct(list(list_of_trues.iloc[i]), list(list_of_preds[i])) for i in range(length)])

## Default

In [9]:
def deadline(timeout, *args):
    def decorate(f):
        def handler(signum, frame):
            raise Exception

        def new_f(*args):
            signal.signal(signal.SIGALRM, handler)
            signal.alarm(timeout)
            return f(*args)
            signal.alarm(0)

        new_f.__name__ = f.__name__
        return new_f
    return decorate

In [10]:
@deadline(500)
def cv_ensemble(model, xs, ys):
    temp = cross_validate(model, xs, ys, return_train_score=True, scoring=make_scorer(average_lowest_correct), n_jobs=1, cv=2)
    return numpy.mean(temp['test_score']).round(2)

In [11]:
cv_ensemble(MultiOutputRegressor(KNeighborsRegressor()), tfidf, ohe)

4.36

## Grid Search

In [12]:
def prn_rsl(gsc):
    prms = gsc.cv_results_['params']
    tsts = (gsc.cv_results_['split0_test_score']+gsc.cv_results_['split1_test_score'])/2
    for i in numpy.argsort(tsts):
        print(prms[i], tsts[i])

In [13]:
ppl = Pipeline([('mb', MultiOutputRegressor(KNeighborsRegressor()))])

In [15]:
grd = {
    'mb__estimator__weights': ['uniform', 'distance'],
    'mb__estimator__algorithm': ['ball_tree', 'kd_tree', 'brute'],
}
gsc = GridSearchCV(estimator=ppl, param_grid=grd, scoring=make_scorer(average_lowest_correct), n_jobs=1, cv=2, verbose=0, error_score=0)
gsc.fit(tfidf, ohe)
prn_rsl(gsc)

{'mb__estimator__algorithm': 'ball_tree', 'mb__estimator__weights': 'distance'} 4.338965638873412
{'mb__estimator__algorithm': 'kd_tree', 'mb__estimator__weights': 'distance'} 4.338965638873412
{'mb__estimator__algorithm': 'brute', 'mb__estimator__weights': 'distance'} 4.338965638873412
{'mb__estimator__algorithm': 'ball_tree', 'mb__estimator__weights': 'uniform'} 4.360682863921491
{'mb__estimator__algorithm': 'kd_tree', 'mb__estimator__weights': 'uniform'} 4.360682863921491
{'mb__estimator__algorithm': 'brute', 'mb__estimator__weights': 'uniform'} 4.360682863921491


In [17]:
grd = {
    'mb__estimator__weights': ['distance'],
#     'mb__estimator__n_neighbors': [i for i in range(10, 100, 10)],
    'mb__estimator__p': [1, 2, 3],
}
gsc = GridSearchCV(estimator=ppl, param_grid=grd, scoring=make_scorer(average_lowest_correct), n_jobs=1, cv=2, verbose=0, error_score=0)
gsc.fit(tfidf, ohe)
prn_rsl(gsc)

{'mb__estimator__p': 2, 'mb__estimator__weights': 'distance'} 4.338965638873412
{'mb__estimator__p': 3, 'mb__estimator__weights': 'distance'} 4.870546466380286
{'mb__estimator__p': 1, 'mb__estimator__weights': 'distance'} 6.971403921850567


In [19]:
grd = {
    'mb__estimator__weights': ['uniform'],
    'mb__estimator__n_neighbors': [i for i in range(10, 100, 10)],
    'mb__estimator__p': [2],
}
gsc = GridSearchCV(estimator=ppl, param_grid=grd, scoring=make_scorer(average_lowest_correct), n_jobs=1, cv=2, verbose=0, error_score=0)
gsc.fit(tfidf, ohe)
prn_rsl(gsc)

{'mb__estimator__n_neighbors': 40, 'mb__estimator__p': 2, 'mb__estimator__weights': 'uniform'} 3.101238833869547
{'mb__estimator__n_neighbors': 70, 'mb__estimator__p': 2, 'mb__estimator__weights': 'uniform'} 3.1033776982943007
{'mb__estimator__n_neighbors': 60, 'mb__estimator__p': 2, 'mb__estimator__weights': 'uniform'} 3.1039258946384596
{'mb__estimator__n_neighbors': 50, 'mb__estimator__p': 2, 'mb__estimator__weights': 'uniform'} 3.121413133346514
{'mb__estimator__n_neighbors': 80, 'mb__estimator__p': 2, 'mb__estimator__weights': 'uniform'} 3.1293642270431548
{'mb__estimator__n_neighbors': 90, 'mb__estimator__p': 2, 'mb__estimator__weights': 'uniform'} 3.140491040135162
{'mb__estimator__n_neighbors': 30, 'mb__estimator__p': 2, 'mb__estimator__weights': 'uniform'} 3.162705910635009
{'mb__estimator__n_neighbors': 20, 'mb__estimator__p': 2, 'mb__estimator__weights': 'uniform'} 3.343397267106422
{'mb__estimator__n_neighbors': 10, 'mb__estimator__p': 2, 'mb__estimator__weights': 'uniform'

### Default
4.36
### Tuned
{'mb__estimator__n_neighbors': 40, 'mb__estimator__p': 2, 'mb__estimator__weights': 'uniform'} 

3.101238833869547